wafer/volunteers/views.py

```diff
@@ -0,0 +1,110 @@
+from django.views.generic import DetailView
+from django.views.generic.list import ListView
+from django.views.generic.edit import UpdateView
+from django.core.urlresolvers import reverse
+from django.http import HttpResponseForbidden
+from django.shortcuts import redirect
+from django.db.models import F, Count
+from django import forms
+
+from wafer.users.views import EditOneselfMixin
+from wafer.volunteers.models import Volunteer, Task
+
+
+class TasksView(ListView):
+    model = Task
+    template_name = 'wafer.volunteers/tasks.html'
+
+    def get_queryset(self):
+        return Task.objects.annotate(nbr_volunteers=Count('volunteers')) \
+                           .filter(nbr_volunteers__lt=F('nbr_volunteers_max'))
+
+    def get_context_data(self, **kwargs):
+        context = super(TasksView, self).get_context_data(**kwargs)
+
+        if self.request.user.is_authenticated():
+            try:
+                volunteer = Volunteer.objects.get(user=self.request.user)
+                context['preferred_tasks'] = Task.objects.annotate(
+                    nbr_volunteers=Count('volunteers')
+                ).filter(
+                    category__in=volunteer.preferred_categories.all(),
+                    nbr_volunteers__lt=F('nbr_volunteers_max')
+                )
+            except Volunteer.DoesNotExist:
+                pass
+
+        return context
+
+
+class TaskView(DetailView):
+    model = Task
+    template_name = 'wafer.volunteers/task.html'
+
+    def get_context_data(self, **kwargs):
+        context = super(TaskView, self).get_context_data(**kwargs)
+        # TODO Find a better way
+        context['already_volunteer'] = (
+            self.request.user.is_authenticated() and
+            self.object.volunteers.filter(user=self.request.user).exists()
+        )
+        context['can_volunteer'] = (
+            self.request.user.is_authenticated() and
+            self.object.nbr_volunteers() < self.object.nbr_volunteers_max
+        )
+        return context
+
+    def post(self, request, *args, **kwargs):
+        if not request.user.is_authenticated():
+            return HttpResponseForbidden()
+        self.object = self.get_object()
+        volunteer, new = Volunteer.objects.get_or_create(user=request.user)
+
+        if self.object in volunteer.tasks.all():
+            volunteer.tasks.remove(self.object)
+            self.object.volunteers.remove(volunteer)
+        elif self.object.nbr_volunteers() < self.object.nbr_volunteers_max:
+            volunteer.tasks.add(self.object)
+            self.object.volunteers.add(volunteer)
+
+        return redirect('wafer_task', pk=self.object.pk)
+
```

Comment:
Is it useful to have the tasks viewable without authenticating?

If it is, should the tasks be included in the static site generation?